### DISCUSSION TITLE
Inbuilt legacy provider

### DISCUSSION THREAD

**Discussion by aniljadaun** [openssl/openssl#21410] [Q&A] [answered]
Hi 

Is it possible to create in-built legacy provider with openssl 3.0 static libraries for windows ?

**Comment by mattcaswell** [comment, marked as answer]
Configure with `no-shared no-module`, e.g.

````
perl Configure no-shared no-module
````

**Reply by aniljadaun** [reply]
Hi @mattcaswell 

Thanks for your suggestion. I did as you mentioned. 
I am creating cipher using symmetric algorithm rc2 and it is working fine with openssl 1.0.2zg though with 3.0 it doesn't work anymore.

I am expecting with inbuilt legacy provider i don't need to load legacy provider separately via code or openssl.cnf file.Working code should not require any change
Am i wrong with my understanding ?

Working code with openssl 1.0.2zg is 

inEncrypt = false;
	if (success && EVP_CipherInit_ex(ctx, algInfo->fAlgID, NULL, NULL, NULL, inEncrypt) != 1)
		success = false;

**Reply by t8m** [reply]
No, even if the legacy provider is built-in, you still have to load it via OSSL_PROVIDER_load() or via the config file.

**Reply by aniljadaun** [reply]
Hi @t8m 

Thanks for your suggestion it worked fine for me.

Any preference when should OSSL_PROVIDER_load(NULL,"legacy") should be done at the. At the init of crypto operation or just before using any legacy algorithm ?

And would it be fine if i generate static libs with legacy module inbuilt from 3.1.x version and use fips module from 3.0.8 version as it is the latest validated fips module. ?

**Reply by mattcaswell** [reply]
Load the legacy provider once during application initialisation.

**Reply by t8m** [reply]
> And would it be fine if i generate static libs with legacy module inbuilt from 3.1.x version and use fips module from 3.0.8 version as it is the latest validated fips module. ?

Yes, that is fine. Of course using the legacy algorithms is non-compliant with FIPS.

**Reply by aniljadaun** [reply]
Hi @t8m legacy module is getting loaded when i created openssl libs with inbuilt legacy module. Though now my fips.dll is not getting loaded. i am using same process.

Error i am getting is : A86C0000:error:07880025:common libcrypto routines:provider_init:reason(37):crypto\provider_core.c:914:name=fips

**Reply by mattcaswell** [reply]
This means it is unable to find the fips.dll. Have you installed it in the appropriate modules directory? BTW you can query the location of the modules directory from your application via `const char *modulesdir = OpenSSL_version(OPENSSL_MODULES_DIR);`. Alternatively you can set the `OPENSSL_CONF` environment variable to point to the location of this directory.

**Reply by aniljadaun** [reply]
Hi @mattcaswell 

	OSSL_LIB_CTX_load_config(NULL, "D:\\outlibsopenssl\\SSL\\openssl.cnf")) 
	const char* default_search_path = "D:\\fipsdll";
	OSSL_PROVIDER_set_default_search_path(NULL, default_search_path);
	fips = OSSL_PROVIDER_load(NULL, "fips");
	
Above code works for me if i use openssl libs which has module created as fips.dll and legacy.dll .separate from libcrypto and libssl . 

In the current scenario i am using inbuilt legacy module and using the  fips module created previously . fips module is not getting loaded for me .
const char *modulesdir = OpenSSL_version(OPENSSL_MODULES_DIR);.
This references to the path on which inbuilt legacy module was created. " D:\opensslnolegacy\Lib\x64\Release\lib\ossl-modules"

Error is same as before
MODULESDIR: "D:\opensslnolegacy\Lib\x64\Release\lib\ossl-modules"
A0330000:error:07880025:common libcrypto routines:provider_init:reason(37):crypto\provider_core.c:914:name=fips

**Reply by aniljadaun** [reply]
creating the inbuilt legacy module didn't generate any fips module separately , so i am using the fips.dll which i created previously.

I also tried adding fips.dll to the path referneced in openssl_modules_dir also added environment as specified by you by creating environment variable. But modulesdir still refrences the same folder

**Reply by mattcaswell** [reply]
> Alternatively you can set the OPENSSL_CONF environment variable to point to the location of this directory.

Sorry - I gave you the wrong environment variable name. The environment variable to use is `OPENSSL_MODULES`. Set it to the name of the directory that contains fips.dll.

**Reply by aniljadaun** [reply]
Hi @mattcaswell 

Unfortunately this is not working for us. Though just replacing with openssl libs with no inbuilt legacy module works. This looks like a bug to me .Can you please check at your end , if this works for you ? i have shared details at https://github.com/openssl/openssl/discussions/21444 . We can close this one and track it from other discussion.

Thanks alot for your help and guidance till now. :)
